### PR TITLE
Add a check for a unique e-mail address

### LINKED ISSUE #440: Lack of email uniqueness check
**To Reproduce**
1. Go to 'https://dev.devadventcalendar.pl/Account/Register'.
2. Register new user (they'll be an error due to wrong EmailSender config, but the user will be created).
3. Register new user again with the previous one email.

**Current behavior**
Currently, only the username is checked for duplication. Multiple users can be registered using the same email address.

**Expected behavior**
The email should be also checked for uniqueness.

**Screenshots**
image

### PR DESCRIPTION
**Pull request by Mejtika**
## Description
Change identity settings to require a unique email address.

## Where should the reviewer start?

Try to register a new user with an existing user name or email.
## Motivation and context
Fix #440 

## How has this been tested?
Manually. Currently, we're checking for email as well as user name uniqueness. 
The screenshots contain errors in the application and the message that will be shown to the user.
image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by WTobor** [comment]
@Mejtika can we test that?

**Comment by Mejtika** [comment]
@WTobor I'll try

**Review by smiechuskce — APPROVED** [review]

File: src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/IntegrationTests/TestStartup/StartupTestBase.cs [hunk 1]
```diff
@@ -21,7 +21,10 @@ public static class StartupTestBase
         static StartupTestBase()
 #pragma warning restore CA1810 // Initialize reference type static fields inline
         {
+            var pathToAppsettingsDev = Path.GetFullPath(@"../../../../DevAdventCalendarCompetition");
```
Comment (WTobor) [inline src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/IntegrationTests/TestStartup/StartupTestBase.cs:24]: we want to add a referenced file to the test folder. In VS - option `Add existing item`
so this logic is not needed. We will have appsettings.Development.json in the test folder. And we need appsettings.Tests.json file (our custom file, only with DefaultConnection).

and logic here should be only
```
.AddJsonFile("appsettings.Development.json")
.AddJsonFile("appsettings.Tests.json"
```

**Comment by WTobor** [comment]
what about renaming appsettings.json to appsettings.test.json?

@patrikulus @Mejtika what about failing LocalDB?

**Comment by Mejtika** [comment]
@WTobor I thought about that, but that additional word after dot by default reflects the environment. Do we have a test env alongside development and production?

**Comment by smiechuskce** [comment]
@WTobor @Mejtika Why do we need a separate Startup test base? I think we shouldn't test the behavior of ASP.NET Identity (assuming that's the reason of adding test test base) but a controller itself. Wouldn't it be enough just to mock the UserManager methods (or a CreateAsync from AccountService) to return failed result in case of adding a user that already exists? That's in fact what I want to do with AccountController and ManageController. Please correct me if I'm wrong.

**Comment by WTobor** [comment]
> @WTobor I thought about that, but that additional word after dot by default reflects the environment. Do we have a test env alongside development and production?

you can set appsettings.tests.json or appsettings.integration-tests.json - you are the author. And yes, it is important to set the environment. As you can see in the repository, we don't have appsettings.test.json. We don't even have appsettings.uat.json (that is our test environment).
I will set appsettins.integration-tests.json name - and we should close the topic of file naming and focus on more important staff ;)

**Comment by WTobor** [comment]
@smiechuskce separate Startup is

> Why do we need a separate Startup test base?

please read about Respawn - that is why we need that class

**Comment by Mejtika** [comment]
@smiechuskce We should think about what we're really wanna test. It's best practice to make sure the test verifies the end result the SUT delivers (black-box): its observable behavior, not the steps it takes to do that. By stubbing the CreateAsync method of AccountService, we're actually testing that our stub works correctly, but the real behavior may be completely different. You've seen, writing test in #428, that that whole mocking/stubbing of identity (UserManager methods) is not so trivial, and such diving into its internals can be misleading. If you even would write that test, would you be sure that it wouldn't lead to a false negative? That PR showed, that we forgot about that line ` config.User.RequireUniqueEmail = true;`. Added tests are checking the hearth of the Register action in Account controller - `_accountService.CreateAsync()`.  Now, we're testing AccountService class that was changed by us and our configuration of ASP.NET Identity, and by using real dependencies we can be sure that it will reflect the real behavior.

**Comment by smiechuskce** [comment]
@Mejtika Ok...now I see the reason. Previously I tried to test the IUserValidator itself (hence all of those internals-digging work) but your solution is much more clear and complete to treat everything as a SUT. Mocking/stubing a CreateAsync would check only the reaction of our controller's Register method and you're right - this could lead to false negative.

Thank you very much for explaining that. I think I should follow this way to write similar test for a ManageController's Index method (POST) in case a trial of changing the name by the user to the existing one.

**Comment by WTobor** [comment]
@Mejtika please resolve conflicts, remove unused appsettings.integration-tests.json (and loading it in Startup) and should be ready

**Comment by WTobor** [comment]
@Mejtika ?

**Review by WTobor — APPROVED** [review]

**Pull request merged**